addons/sale/migrations/9.0.1.0/post-migration.py

```diff
@@ -169,10 +168,52 @@ def set_so_line_computed_rest(env):
         WHERE so.id = sol.order_id""")
 
 
+def custom_clean_refund_order_line_rel(env):
+    """ OD12-496: delete the order line / invoice line relation
+    between refunds of service products in those cases where they
+    would cause the line to get invoice_status 'to invoice'
+    """
+    openupgrade.logged_query(
+        env.cr,
+        """
+WITH qties AS (
+    SELECT sol.id, sol.order_id, sol.product_uom_qty,
+        ROUND(SUM(ail.quantity / u.factor * u2.factor), 2) as qinvoiced
+    FROM sale_order_line sol
+    JOIN sale_order_line_invoice_rel rel ON rel.order_line_id = sol.id
+    JOIN account_invoice_line ail ON ail.id = rel.invoice_line_id
+    JOIN account_invoice ai ON ai.id = ail.invoice_id
+    JOIN product_product pp ON pp.id = sol.product_id
+    JOIN product_template pt ON pt.id = pp.product_tmpl_id
+    JOIN product_uom u on (u.id = ail.uom_id)
+    JOIN product_uom u2 on (u2.id = sol.product_uom)
+    WHERE pt.type = 'service'  -- pt.invoice_policy = 'order', but see OD12-491
+        AND ai.type = 'out_invoice' AND ai.state != 'cancel'
+        AND EXISTS (
+            SELECT * FROM sale_order_line_invoice_rel rel2
+            JOIN account_invoice_line ail2 ON ail2.id = rel2.invoice_line_id
+            JOIN account_invoice ai2 ON ai2.id = ail2.invoice_id
+            WHERE rel2.order_line_id = rel.order_line_id
+                AND ai2.type = 'out_refund' AND ai2.state != 'cancel')
+    GROUP BY sol.id, sol.product_uom_qty
+    HAVING product_uom_qty >= ROUND(SUM(ail.quantity / u.factor * u2.factor), 2)
+    )
+DELETE FROM sale_order_line_invoice_rel
+WHERE invoice_line_id IN (
+    SELECT ail.id FROM account_invoice_line ail
+    JOIN sale_order_line_invoice_rel rel ON rel.invoice_line_id = ail.id
+    JOIN qties ON qties.id = rel.order_line_id
+    JOIN account_invoice ai ON ai.id = ail.invoice_id
+    WHERE ai.type = 'out_refund' AND ai.state != 'cancel'
+    );
+        """)
```

Comment:
do all these invoices have the the good origin field, or how do we know that those invoices still belong to their orders?

Reply:
I checked that all these invoices were linked to the sale order through sale_order_invoice_rel, so yes.

Reply:
but we don't use sale_order_invoice_rel any more, or is the idea that I revert https://github.com/VanMoof/sale-financial/pull/3 to looking at this field too?

Reply:
sale_order_invoice_rel serves as the basis for the invoices' origin updates here: https://github.com/VanMoof/odoo-addons/blob/12.0/scripts/20200326-migrate_all.bash#L133-L136

Reply:
clear